Executing Vincent Tools Doc (#141)
LIT-Protocol/Vincent: packages/libs/app-sdk/docs/Developers/App-Agent-Developers/Executing-Tools.md

---
category: Developers
title: Executing Vincent Tools
---

# Using the Vincent Tool Client to Execute Tools

> As a Vincent App Developer, you need a structured way to execute the Vincent Tools your App Users have permitted you the ability to execute on their behalf.

Reviewer: Maybe consider
```suggestion
As a Vincent App Developer, you need a structured way to execute Vincent Tools for your Vincent App Users.
```

   - Returns the execution results of both the Tool and any evaluated Policies

# Creating a Tool Client

To create an instance of the Tool Client, import the `getVincentToolClient` function from the `@lit-protocol/vincent-app-sdk` package. This function takes two parameters: a Vincent Tool definition and an ethers signer, returning an instance of the Tool Client.

The following code uses an example Vincent Tool package for reference:

```typescript
import { ethers } from 'ethers';
import { getVincentToolClient } from '@lit-protocol/vincent-app-sdk';
import { bundledVincentTool } from '@example-org/vincent-tool-example-erc20-transfer';

const ethersSigner = new ethers.Wallet(process.env.VINCENT_APP_DELEGATEE_PRIVATE_KEY);

const toolClient = getVincentToolClient({
  bundledVincentTool,
  ethersSigner,
});
```

The two required parameters for the `getVincentToolClient` function are:

1. `bundledVincentTool`: The definition of the Vincent Tool you want to execute on behalf of the App User, imported from a Vincent Tool package

>    - This tool definition is exported by the author of the Vincent Tool package you are using and defines properties like the expected input parameters of the Tool, the Vincent Policies supported by the Tool, and the Tool's expected return values

Reviewer: Consider
```suggestion
   - This tool definition is exported by the author of the Vincent Tool package and defines properties like the expected input parameters of the Tool, the Vincent Policies supported by the Tool, and the Tool's expected return values
```

If `precheck` returns a success result, you can execute the Tool's `execute` function with reasonable confidence that the Tool should execute successfully.

<!-- TODO Link to PrecheckFailSchema typedoc interface when it's available -->

However, if one of the evaluated Policies returns a deny response, signifying it's expecting to deny Tool execution, or all Policies permit execution, but the Tool's `precheck` function is expecting Tool execution to fail, the `precheck` function will return a `ToolResponse` with a failure result that has the structure of the Tool's Zod schema: `PrecheckFailSchema`.

If `precheck` returns a failure result, you should check the `error` property on the result object to see if there was a known error that occurred during the execution of the Tool's `precheck` function. Additionally, if one of the evaluated Polices has denied Tool execution, you should check the result object's `context.policiesContext` property to see which Policy denied Tool execution and why.

# Executing the Tool Client's `execute` function

<div class="info-box">
  <p class="info-box-title">
    <span class="info-icon">Info</span> Before executing the Tool
  </p>

>   <p>Executing a Tool using the Lit Protocol network requires a <a href="https://developer.litprotocol.com/paying-for-lit/capacity-credits">Lit Capacity Credit</a> minted for the Ethereum Address you're using for the `ethersSigner` when creating the Tool Client instance with `getVincentToolClient`.</p>

Reviewer: Not a huge deal, but the Markdown on `ethersSigner` and `getVincentToolClient` doesn't seem to be applied on the deployment

Author: Ah didn't notice that. Probably have to make a class to style it since it's HTML and not markdown - Will look into it
